core/src/main/java/com/linecorp/armeria/server/encoding/HttpDecodedRequest.java

```diff
@@ -59,7 +61,16 @@ protected void beforeComplete(Subscriber<? super HttpObject> subscriber) {
 
     @Override
     protected Throwable beforeError(Subscriber<? super HttpObject> subscriber, Throwable cause) {
+        if (cause instanceof CancelledSubscriptionException) {
+            // We already handle it in beforeCancel().
+            return cause;
+        }
         responseDecoder.finish();
         return cause;
     }
+
+    @Override
+    protected void beforeCancel(Subscriber<? super HttpObject> subscriber, Subscription subscription) {
+        responseDecoder.finish();
```

Comment:
If responseDecoder.finish() is idempotent, we could remove `cause instanceof CancelledSubscriptionException` above.